app/scripts/controllers/plugins.js

```diff
@@ -0,0 +1,149 @@
+    this._onNewTx = opts._onNewTx
+    this._subscribeToPreferencesControllerChanges = opts._subscribeToPreferencesControllerChanges
+    this._updatePreferencesControllerState = opts._updatePreferencesControllerState
+    this._signPersonalMessage = opts._signPersonalMessage
+    this._getAccounts = opts._getAccounts
+  }
+
+  get (pluginName) {
+    return this.store.getState().plugins[pluginName]
+  }
+
+  // When a plugin is first created, where should it be executed?
+    // And how do we ensure that the same plugin is never executed twice?
+
+  updatePluginState (pluginName, newPluginState) {
+    const plugins = this.store.getState().plugins
+    const plugin = plugins[pluginName]
+    const updatedPlugin = { ...plugin, pluginState: newPluginState }
+
+    const newPlugins = {...plugins, [pluginName]: updatedPlugin}
+
+    this.store.updateState({
+      plugins: newPlugins,
+    })
+  }
+
+  async create (pluginName) {
+    const plugins = this.store.getState().plugins
+
+    // if (plugins[pluginName]) {
+    //   return plugins[pluginName]
+    // }
+
+    const { source, uiWrappers, requestedAPIs } = await this._getPluginConfig(pluginName)
+    const accounts = await this._getAccounts()
+    console.log('!!!!!* p accounts', accounts)
+    const address = accounts[0]
+
+    const newPlugin = {
+      handleRpcRequest: async (result) => {
+        return Promise.resolve(result)
+      },
+      pluginName,
+      source,
+      pluginState: {},
+      uiWrappers,
+    }
+
+    const newPlugins = {...plugins, [pluginName]: newPlugin}
+
+    this.store.updateState({
+      plugins: newPlugins,
+    })
+
+    this._startPlugin(pluginName, source, requestedAPIs, address)
+
+    return newPlugin
+  }
+
+  _generateApisToProvide (apiList, pluginName) {
+    const updatePluginState = this.updatePluginState.bind(this, pluginName)
+    const possibleApis = {
+      fetch,
+      updatePluginState,
+      onNewTx: this._onNewTx,
+      onUnlock: this._onUnlock,
+      Box,
+      subscribeToPreferencesControllerChanges: this._subscribeToPreferencesControllerChanges,
+      updatePreferencesControllerState: this._updatePreferencesControllerState,
+      generateSignature: (text, address) => {
+        const msg = '0x' + Buffer.from(text, 'utf8').toString('hex')
+
+        return this._signPersonalMessage({
+          from: address,
+          data: msg,
+        })
+      }
+    }
+    const apisToProvide = {}
+    apiList.forEach(apiKey => {
+      apisToProvide[apiKey] = possibleApis[apiKey]
+    })
+    return apisToProvide
+  }
+
+  _startPlugin (pluginName, source, requestedAPIs, address) {
+    const s = SES.makeSESRootRealm({consoleMode: 'allow', errorStackMode: 'allow', mathRandomMode: 'allow'})
+    const newPluginSessified = s.evaluate(source, {
+      pluginAPIs: this._generateApisToProvide(requestedAPIs, pluginName),
+      pluginData: {
+        address,
+      },
+      ethereumProvider: this.provider,
```

Comment:
Each plugin must get a fresh provider restricted to its own domain. (Maybe not for MVP, but must for security).

Made much easier with [this pending PR](https://github.com/MetaMask/metamask-extension/pull/6670).

Reply:
Thanks. I hadn't considered that yet. I will follow that PR, rebase and update this one accordingly.